### ISSUE TITLE
BluRay m2ts failure for linux 64-bit tsMuxeR

### ISSUE THREAD

**Issue by chadady** [justdan96/tsMuxer#316]
Compiled for 32-bit works fine.  The GUI will report error 254.  This is the CLI output:

linuxbox:/dc$ ~/tsMuxer/bin/tsMuxeR Terminator_Genisys.meta Terminator_Genisys.m2ts
tsMuxeR version git-c61c002. github.com/justdan96/tsMuxer
Decoding H264 stream (track 1): Profile: High@4.1  Resolution: 1920:1080p  Frame rate: 23.976
H.264 muxing fps is not set. Get fps from stream. Value: 23.976
0.0% complete
Decoding TRUE-HD stream (track 2): AC3 core+TRUE-HD. Peak bitrate: 7878Kbps (core 640Kbps) Sample Rate: 48KHz Channels: 7.1
TRUE-HD stream (track 2): bad frame detected at position00:00:00,544. Resync stream.
Decoding TRUE-HD stream (track 2): AC3 core+TRUE-HD. Peak bitrate: 7878Kbps (core 640Kbps) Sample Rate: 48KHz Channels: 7.1
B-pyramid level 2 detected. Shift DTS to 3 frames
0.7% complete
Decoding PGS stream (track 3):  Resolution: 1920:1080  Frame rate: 23.976
0.9% complete
TRUE-HD stream (track 2): bad frame detected at position00:01:13,408. Resync stream.
Decoding TRUE-HD stream (track 2): AC3 core+TRUE-HD. Peak bitrate: 7878Kbps (core 640Kbps) Sample Rate: 48KHz Channels: 7.1
1.3% complete

:
:
:
Decoding TRUE-HD stream (track 2): AC3 core+TRUE-HD. Peak bitrate: 7878Kbps (core 640Kbps) Sample Rate: 48KHz Channels: 7.1
10.5% complete
TRUE-HD stream (track 2): bad frame detected at position00:12:27,360. Resync stream.
Decoding TRUE-HD stream (track 2): AC3 core+TRUE-HD. Peak bitrate: 7878Kbps (core 640Kbps) Sample Rate: 48KHz Channels: 7.1
TRUE-HD stream (track 2): bad frame detected at position00:12:27,552. Resync stream.
Decoding TRUE-HD stream (track 2): AC3 core+TRUE-HD. Peak bitrate: 7878Kbps (core 640Kbps) Sample Rate: 48KHz Channels: 7.1
Reading buffer overflow. Possible container streams are not syncronized. Please, verify stream fps. File name: /dc/Videos/backup/Terminator_Genisys/BDMV/STREAM/00800.m2ts

The meta file:

MUXOPT --no-pcr-on-video-pid --new-audio-pes --hdmv-descriptors --vbr  --vbv-len=500 --start-time=524280
V_MPEG4/ISO/AVC, "/dc/Videos/backup/Terminator_Genisys/BDMV/PLAYLIST/00800.mpls", insertSEI, contSPS, track=4113
A_AC3, "/dc/Videos/backup/Terminator_Genisys/BDMV/PLAYLIST/00800.mpls", track=4352, lang=eng
#A_AC3, "/dc/Videos/backup/Terminator_Genisys/BDMV/PLAYLIST/00800.mpls", track=4353, lang=fra
#A_AC3, "/dc/Videos/backup/Terminator_Genisys/BDMV/PLAYLIST/00800.mpls", track=4354, lang=spa
#A_AC3, "/dc/Videos/backup/Terminator_Genisys/BDMV/PLAYLIST/00800.mpls", track=4355, lang=por
#A_AC3, "/dc/Videos/backup/Terminator_Genisys/BDMV/PLAYLIST/00800.mpls", track=4356, lang=eng
S_HDMV/PGS, "/dc/Videos/backup/Terminator_Genisys/BDMV/PLAYLIST/00800.mpls", fps=23.976, track=4608, lang=eng
#S_HDMV/PGS, "/dc/Videos/backup/Terminator_Genisys/BDMV/PLAYLIST/00800.mpls", fps=23.976, track=4609, lang=eng
#S_HDMV/PGS, "/dc/Videos/backup/Terminator_Genisys/BDMV/PLAYLIST/00800.mpls", fps=23.976, track=4610, lang=fra
#S_HDMV/PGS, "/dc/Videos/backup/Terminator_Genisys/BDMV/PLAYLIST/00800.mpls", fps=23.976, track=4611, lang=spa
#S_HDMV/PGS, "/dc/Videos/backup/Terminator_Genisys/BDMV/PLAYLIST/00800.mpls", fps=23.976, track=4612, lang=por

I've verifed the failure is also seen on the nightly build.  I'm running Ubuntu 20.04.

If I add these lines to the tsmuxer and libmediation CMakeLists.txt files, forcing a 32-bit compile, all is well.

set(CMAKE_CXX_FLAGS -m32)
set(CMAKE_EXE_LINKER_FLAGS -m32)

The failure was introduced with commit 1299e968 - Merge branch 'master' into master.  This commit turned on 64-bit.

The same failure is seen if I use the GUI to select the ac3 core. Explicitly specifying the frame rate does not fix it either.  I suspect the original code is not 64-bit safe.

On a side note, to build on Ubuntu 20.04, the 64-bit compile notes should be updated:

the ninja package is named ninja-build for Ubuntu
qttools5-dev also must be installed to build the GUI

Hope this helps.

Thanks for making this available and maintaining this project!

**Comment by xavery** [comment]
> I suspect the original code is not 64-bit safe.

It certainly wasn't and we can't really be 100% sure that it is now. A few fixes were needed to even get it to compile, and a few more to stop it from crashing once it did. I suspect you might've hit one of the many places where unplanned truncation happens or a (possibly indirectly) expected overflow condition doesn't happen because of the extra width.

I'm afraid this can't be investigated further without a sample that reproduces this issue.

**Comment by chadady** [comment]
I found the problem and can provide a patch.

[0001-Use-memmove-when-buffers-can-overlap.patch.gz](https://github.com/justdan96/tsMuxer/files/5079765/0001-Use-memmove-when-buffers-can-overlap.patch.gz)

**Comment by xavery** [comment]
What a great catch. How did you figure out that it might be `memcpy`/`memmove` causing the bug?

**Comment by chadady** [comment]
The tool worked fine compiled 32-bit, so I added a lot of debug and ran it as 64-bit/32-bit to try to understand how data moved around and why it was failing.  I don't know much about C++ so I found it difficult to understand how data was read and then dispatched to the proper reader.  The tool is impressive in how it manages input files, using threads to manage them.  Data is asynchronously read and staged in buffers that are then drained by 'readers'.  The readers grab a chunk of data and process it in a temporary buffer.  As data is consumed it is 'resynced' in that buffer or another as needed.  The resync is implemented with memcpy() moving data within the same buffer (overlap).

I spent a lot of time working the wrong assumption that it was a pointer issue.  In the end I started running adler-32 data checks whenever a data block was moved or modified to see when there was a mismatch which narrowed it down to the memcpy().  I went to adler-32 from CRC because I couldn't see how that code broke, being convinced that the data corruption was a block shift in a buffer due to a pointer error.

I had forgotten that memcpy() did not support overlapping buffers.  A question to ask is 'why does this work in 32-bit?'.  I believe that back in the 32-bit days memcpy() had no such restriction, and this came later.  I seem to recall some grousing from the community when this change came so my guess is 32-bit memcpy() works here.  It could also simply be that we've been lucky.

I also don't understand why this has not been reported before.  I've tried two different blu-ray data sets and both fail.  The 64-bit update was from August of last year.  I'm hopeful that this bug will fix other outstanding issues as well.
